crates/libs/core/src/interface.rs

```diff
@@ -98,17 +119,93 @@ pub unsafe trait Interface: Sized + Clone {
     /// # Safety
     ///
     /// `interface` must be a non-null, valid pointer for writing an interface pointer.
-    unsafe fn query(&self, iid: *const GUID, interface: *mut *mut std::ffi::c_void) -> HRESULT {
+    #[inline(always)]
+    unsafe fn query(&self, iid: *const GUID, interface: *mut *mut c_void) -> HRESULT {
         if Self::UNKNOWN {
             (self.assume_vtable::<IUnknown>().QueryInterface)(self.as_raw(), iid, interface)
         } else {
             panic!("Non-COM interfaces cannot be queried.")
         }
     }
+
+    /// Creates an `InterfaceRef` for this reference. The `InterfaceRef` tracks lifetimes statically,
+    /// and eliminates the need for dynamic reference count adjustments (AddRef/Release).
+    fn to_ref(&self) -> InterfaceRef<'_, Self> {
+        InterfaceRef::from_interface(self)
+    }
 }
 
 /// # Safety
 #[doc(hidden)]
-pub unsafe fn from_raw_borrowed<T: Interface>(raw: &*mut std::ffi::c_void) -> Option<&T> {
+pub unsafe fn from_raw_borrowed<T: Interface>(raw: &*mut c_void) -> Option<&T> {
     T::from_raw_borrowed(raw)
 }
+
+/// This has the same memory representation as `IFoo`, but represents a borrowed interface pointer.
+///
+/// This type has no `Drop` impl; it does not AddRef/Release the given interface. However, because
+/// it has a lifetime parameter, it always represents a non-null pointer to an interface.
+#[repr(transparent)]
+pub struct InterfaceRef<'a, I>(NonNull<c_void>, PhantomData<&'a I>);
+
+impl<'a, I> Copy for InterfaceRef<'a, I> {}
```

Comment:
Derive these traits?

Reply:
Can't.  `#[derive]` does the wrong thing with generic types.  It requires that `T` be `Copy`, which is wrong.